src/IO/WriteBufferFromS3.cpp

```diff
@@ -700,12 +697,11 @@ void WriteBufferFromS3::makeSinglepartUpload(WriteBufferFromS3::PartData && data
             if (client_ptr->isClientForDisk())
                 ProfileEvents::increment(ProfileEvents::DiskS3PutObject);
 
-            ResourceCost cost = request.GetContentLength();
-            ResourceGuard rlock(write_settings.resource_link, cost);
+            CurrentThread::IOScope io_scope(write_settings.io_scheduling);
```

Comment:
Here you are controlling only particular current tread with the exact call. In other words you enable scheduling only for PUT (single object + multi parts) and GET (get object).
Also you manually set resources guards for Azure and HDFS.

But what about the rest? I expected that you cover here HTTP storages as well with other types of requests.

Reply:
Yes, it should cover all remote and local IO eventually. In this PR I'm just trying to improve what we already have. Let's also do the following in a separate PR:
1. move `CurrentThread::IOScope` into  ObjectStorage
2. make IOScope inheritable when spawning new threads 
3. get rid of `ReadSettings` and `WriteSettings` fields for passing `io_scheduling` field
It should make the process of integration other buffers and scheduler more straightforward